src/main/kotlin/app/CodeLongevity.kt

```diff
@@ -0,0 +1,173 @@
+import org.eclipse.jgit.diff.RawText
+import org.eclipse.jgit.internal.storage.file.FileRepository
+import org.eclipse.jgit.lib.Repository
+import org.eclipse.jgit.revwalk.RevCommit
+import org.eclipse.jgit.revwalk.RevWalk
+import org.eclipse.jgit.revwalk.RevSort
+import org.eclipse.jgit.treewalk.filter.*
+import org.eclipse.jgit.treewalk.TreeWalk
+import org.eclipse.jgit.util.io.DisabledOutputStream
+
+import java.util.Date
+
+/**
+ * Used to store and identify a code line information in a repo.
+ */
+class CodeLine(val rev: String, val file: String, val line: Int,
+               val authorName: String, val authorEmail: String,
+               val age: Int) {
+}
+
+/**
+ * Used to compute age of code lines in the repo.
+ */
+class CodeLongevity(repo_path: String) {
+    var repo_path: String = repo_path
+
+    var list: MutableList<CodeLine> = mutableListOf()
+
+    init {
+        println("READYY>>>>>>>>>>");
+        println("Repo: $repo_path")
+
+        val repo = FileRepository(repo_path)
+        val revWalk = RevWalk(repo);
+        val head = revWalk.parseCommit(repo.readOrigHead()); // master
+
+        computeExistingLines(repo, head)
+        computeDeletedLines(repo, revWalk, head)
+    }
+
+    fun test(email: String) {
+        var sum: Long = 0;
+        for (line in list) {
+            if (line.authorEmail != email) {
+                continue
+            }
+            sum += line.age
+        }
+        if (list.size > 0) {
+            println("avg code line age for <$email> is ${sum / list.size} seconds")
+        }
+    }
+
+    /**
+     * Computes age of all existing lines in all files of the repo.
+     */
+    private fun computeExistingLines(repo: FileRepository, head: RevCommit) {
+        val treeWalk = TreeWalk(repo);
+        treeWalk.setRecursive(true);
+        treeWalk.addTree(head.getTree());
+        while (treeWalk.next()) {
+            // todo: skip binary files.
+
+            val file = treeWalk.getPathString();
+            //println("found: ${file}");
+            val blameGen = BlameGenerator(repo, file)
+            blameGen.push(null, head)
+            val blame = blameGen.computeBlameResult()
+            val lines = blame.getResultContents()
+            var index = 0;
+            while (index < lines.size()) {
+                val blameCommit = blame.getSourceCommit(index)
+                val timeDiff = (head.getCommitTime() - blameCommit.getCommitTime())
+                val author = blame.getSourceAuthor(index)
+                val authorName = author.getName()
+                val authorEmail = author.getEmailAddress()
+
+                list.add(CodeLine(blameCommit.getName(),
+                                  file, index,
+                                  authorName, authorEmail, timeDiff))
+                index++
+            }
+        }
+    }
+
+    /**
+     * Computes age of all deleted lines in the repo.
+     */
+    private fun computeDeletedLines(repo: FileRepository,
```

Comment:
Here is what worries me in this approach: (i) it potentially traverses repository more than once, (ii) it makes a counter-intuitive separation of lines that are present in head vs lines that were at some point deleted, (iii) it is not trivial to make it incremental.

Why not something cruder but possibly cleaner, e.g.:
```
- Maintain a hashmap: path -> list_of_lines, where each item in the list of lines is
   a couple (author, time)
- Iterate from the initial commit to head in the master (direct chronological order):
    - Iterate over all blobs in the commit:
        - if blob path is not in hash_map:
            - add a new entry: blob_path -> (commit_author, commit_date) *  N times,
               where N is the number of lines in the blob.
        - else:
            - for all deleted lines, emit (author, commit_time - time)
            - for all added lines, insert (commit_author, commit_time) into appropriate
               location in the list
```
Then we can save the arrays emitted during this operation, and simply restart the process once new commits appear on top of master.

We could also decide what to do with lines that were not overwritten yet. Those potentially have been living for a long time, so it would be unfair to discard them, so we can do something like (now - time) * 2, to give everybody some benefit of doubt and to recognize their lines are still alive in the latest and greatest version of the repo.